### PR TITLE
Blackbox rr

### PR DESCRIPTION
**Pull request by rdaly525**

### CODE REVIEW

File: tests/test_black_box_rr.py [hunk 1]
```diff
@@ -51,3 +48,93 @@ def test_fp_pe_smt():
...
+    arch_fc = fp.PE_fc
+    arch_mapper = ArchMapper(arch_fc)
+
+    ir_mapper = arch_mapper.process_ir_instruction(ir_fc, simple_formula=True)
+    rewrite_rule = ir_mapper.solve('z3', external_loop=True)
+    if not found:
+        assert rewrite_rule is None
+        return
+    assert rewrite_rule is not None
+    #verify the mapping works
```
Comment (cdonovick) [inline tests/test_black_box_rr.py:131]: remove commented code

File: tests/test_black_box_rr.py [hunk 2]
```diff
@@ -51,3 +48,93 @@ def test_fp_pe_smt():
...
+    ):
+        ir_mapper = arch_mapper.process_ir_instruction(ir_fc, simple_formula=True)
+        rewrite_rule = ir_mapper.solve('z3', external_loop=True)
+        if found:
+            assert rewrite_rule is not None
+        else:
+            assert rewrite_rule is None
+
+        #verify the mapping works
+        #counter_example = rewrite_rule.verify()
```
Comment (cdonovick) [inline tests/test_black_box_rr.py:105]: remove commented code

File: peak/family.py [hunk 3]
```diff
@@ -193,7 +193,13 @@ def assemble(self, locals, globals, **kwargs):
         s_deco = super().assemble(locals, globals, **kwargs)
         def deco(cls):
             if issubclass(cls, BlackBox):
-                cls.__call__ = BlackBox.__call__
+                old_call = cls.__call__
+
+                cls.__call__ = BlackBox.create_call()
```
Comment (cdonovick) [inline peak/family.py:198]: make create a call a `@classmethod` then just call `cls.create_call()` which does all this in place. e.g.

```Python
@classmethod
def create_call(cls):
    def call(...): ...
    input_t = ...
    cls.__call__ = call
    cls.__call__.input_t = input_t
```
The current solution kinda has a weird intertangling of responsibilities for each function.  
At least move the setting of input_t into `create_call` so this is just `cls.__call__ = BlackBox.create_call()`

**Review by cdonovick — CHANGES_REQUESTED** [review]

File: tests/test_black_box_rr.py (hunk 1 above)
Comment (rdaly525) [reply tests/test_black_box_rr.py]: Fixed

File: tests/test_black_box_rr.py (hunk 2 above)
Comment (rdaly525) [reply tests/test_black_box_rr.py]: Fixed

File: peak/family.py (hunk 3 above)
Comment (rdaly525) [reply peak/family.py]: Fixed

**Review by cdonovick — APPROVED** [review]

**Pull request merged**